### PR TITLE
Just an idea for review: Event dispatcher

### PR DESCRIPTION
**Pull request by gauntface**
cc @wibblymat @jeffposnick @addyosmani 

This is an idea for implementing: https://github.com/GoogleChrome/Propel/issues/4

I've done some rejigging of what happens when a new subscription is obtained, but mostly this is just adding events in.

This is a demo in /test/data/demo/index.html that you can check out how it could be used. If you want to run it, just run:

    node test/server/index.js

### CODE REVIEW

**Comment by wibblymat** [comment]
My major comment at the moment is that this doesn't use the DOM CustomEvent system. I don't know enough about it to know whether it's a good choice, but seems like something worth discussing.

**Comment by wibblymat** [comment]
@gauntface I read up on `CustomEvent` and it totally isn't suitable, so I'm going to review this properly now!

File: src/client/event-dispatch.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+/* eslint-env browser */
+
+'use strict';
+
+export default class EventDispatch {
+  constructor() {
+    this._eventTypes = {};
```
Comment (wibblymat) [inline src/client/event-dispatch.js]: We can be more ES6:

```
this._eventTypes = new Map();

...

addEventListener(type, listener)  {
  if (!this._eventTypes.has(type)) {
    this._eventTypes.set(type, new Set());
  }

  this._eventTypes.get(type).add(listener);
}

removeEventListener(type, listener) {
  if (this._eventTypes.has(type)) {
    this._eventTypes.get(type).delete(listener);
  }
}
```

removes the need for `indexOf`.

**Comment by gauntface** [comment]
@wibblymat this should address the map and sets comment.

@owencm I'll address the constant updating issue in a seperate PR by adding indexdb to track subscription objects.

**Comment by wibblymat** [comment]
:+1:

**Comment by wibblymat** [comment]
Looks good. We can refine the idea in future PRs.

**Pull request merged**